File: pipecat/overwatch_pipeline/auth/token_validator.py
"""Python mirror of the TS daemon's TokenValidator.

The phone derives a per-session token at session-start time as:

    sig = HMAC-SHA256(pairing_token, f"{session_id}|{expires_at}")
    token = f"{session_id}|{expires_at}|{sig.hex()}"

The relay forwards the token verbatim to Pipecat Cloud's session-start API,
which surfaces it to the bot via `runner_args.body.session_token`. The bot
forwards it again on every harness_command envelope; the daemon's
`createTokenValidator` (TS) verifies HMAC + expiry there.

This module exists so the orchestrator can also check the token at its own
boundary — refuse to start the pipeline at all when the token is expired or
signed with the wrong pairing secret. Without this, a tampered token rides
all the way to the daemon before being rejected, leaving the user staring
at a dead Daily room.

Wire format must match TS exactly. See:
- packages/session-host-daemon/src/adapter-protocol/token-validator.ts
- overwatch-mobile/src/services/session-token.ts
- tests/cross-runtime-token-contract.test.ts
"""

from __future__ import annotations

import hmac
import time
from dataclasses import dataclass
from hashlib import sha256
from typing import Protocol


@dataclass(frozen=True)
class SessionTokenClaims:
    session_id: str
    expires_at: int  # unix seconds


class TokenValidator(Protocol):
    def issue(self, claims: SessionTokenClaims) -> str: ...
    def verify(self, token: str) -> SessionTokenClaims | None: ...


def create_token_validator(pairing_token: str) -> TokenValidator:
    if not pairing_token:
        raise ValueError("token-validator: empty pairing token")
    secret = pairing_token.encode("utf-8")

    def _sign(payload: str) -> str:
        return hmac.new(secret, payload.encode("utf-8"), sha256).hexdigest()

    class _Impl:
        def issue(self, claims: SessionTokenClaims) -> str:
            payload = f"{claims.session_id}|{claims.expires_at}"
            return f"{payload}|{_sign(payload)}"

        def verify(self, token: str) -> SessionTokenClaims | None:
            if not isinstance(token, str) or not token:
                return None
            parts = token.split("|")
            if len(parts) != 3:
                return None
            session_id, expires_str, signature = parts
            try:
                expires_at = int(expires_str)
            except ValueError:
                return None

            expected = _sign(f"{session_id}|{expires_at}")
            # hmac.compare_digest is constant-time.
            if not hmac.compare_digest(signature, expected):
                return None

            if expires_at < int(time.time()):
                return None

            return SessionTokenClaims(session_id=session_id, expires_at=expires_at)

    return _Impl()

File: packages/session-host-daemon/src/adapter-protocol/token-validator.ts
/**
 * Validates per-user + per-session tokens on inbound HarnessCommands.
 *
 * The pairing token (long-term, per-user) is bootstrapped via the QR-pair flow
 * during `overwatch setup`. The cloud orchestrator presents a per-session token
 * derived as HMAC(pairing_token, session_id || expires_at). The daemon verifies
 * the HMAC and the expiry.
 */

import { createHmac, timingSafeEqual } from "node:crypto";

export interface SessionTokenClaims {
  session_id: string;
  expires_at: number; // unix seconds
}

export interface TokenValidator {
  /**
   * Issue a per-session token. Used in tests and for daemon-side simulation;
   * in production, the phone derives the token and presents it to the
   * orchestrator who then forwards on commands.
   */
  issue(claims: SessionTokenClaims): string;

  /** Verify a presented session token. Returns claims on success, null on failure. */
  verify(token: string): SessionTokenClaims | null;
}

export function createTokenValidator(pairingToken: string): TokenValidator {
  if (!pairingToken) {
    throw new Error("token-validator: empty pairing token");
  }

  const sign = (payload: string): string =>
    createHmac("sha256", pairingToken).update(payload).digest("hex");

  return {
    issue(claims) {
      const payload = `${claims.session_id}|${claims.expires_at}`;
      const signature = sign(payload);
      return `${payload}|${signature}`;
    },

    verify(token) {
      if (!token || typeof token !== "string") return null;
      const parts = token.split("|");
      if (parts.length !== 3) return null;
      const [session_id, expiresStr, signature] = parts;
      const expires_at = Number.parseInt(expiresStr, 10);
      if (!Number.isFinite(expires_at)) return null;

      const expected = sign(`${session_id}|${expires_at}`);
      const a = Buffer.from(signature, "hex");
      const b = Buffer.from(expected, "hex");
      if (a.length !== b.length || !timingSafeEqual(a, b)) return null;

      const now = Math.floor(Date.now() / 1000);
      if (expires_at < now) return null;

      return { session_id, expires_at };
    },
  };
}
